app/src/main/java/org/wikipedia/analytics/eventplatform/ArticleInteractionEvent.kt

```diff
@@ -0,0 +1,150 @@
+package org.wikipedia.analytics.eventplatform
+
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import org.wikipedia.analytics.eventplatform.ArticleInteractionEvent.ActionType.*
+import org.wikipedia.auth.AccountUtil
+
+@Suppress("unused")
+@Serializable
+@SerialName("/analytics/mobile_apps/android_article_toolbar_interaction/1.0.0")
+class ArticleInteractionEvent(@SerialName("wiki_db") private var wikiDb: String,
+                              @SerialName("page_id") private var pageId: Int) : TimedEvent(STREAM_NAME) {
+
+    @SerialName("time_spent_ms") private var timeSpentMs: Int? = null
+    @SerialName("is_anon") private var isAnon: Boolean? = null
+    private var action: String? = null
+
+    fun logSaveClick() {
+        action = SAVE.valueString
+        submitEvent()
+    }
+
+    fun logLanguageClick() {
+        action = LANGUAGE.valueString
+        submitEvent()
+    }
+
+    fun logFindInArticleClick() {
+        action = FIND_IN_ARTICLE.valueString
+        submitEvent()
+    }
+
+    fun logThemeClick() {
+        action = THEME.valueString
+        submitEvent()
+    }
+
+    fun logContentsClick() {
+        action = CONTENTS.valueString
+        submitEvent()
+    }
+
+    fun logMoreClick() {
+        action = MORE.valueString
+        submitEvent()
+    }
+
+    fun logShareClick() {
+        action = SHARE.valueString
+        submitEvent()
+    }
+
+    fun logTalkPageClick() {
+        action = TALK_PAGE.valueString
+        submitEvent()
+    }
+
+    fun logEditHistoryClick() {
+        action = EDIT_HISTORY.valueString
+        submitEvent()
+    }
+
+    fun logNewTabClick() {
+        action = NEW_TAB.valueString
+        submitEvent()
+    }
+
+    fun logExploreClick() {
+        action = EXPLORE.valueString
+        submitEvent()
+    }
+
+    fun logForwardClick() {
+        action = FORWARD.valueString
+        submitEvent()
+    }
+
+    fun logNotificationClick() {
+        action = NOTIFICATION.valueString
+        submitEvent()
+    }
+
+    fun logTabsClick() {
+        action = TABS.valueString
+        submitEvent()
+    }
+
+    fun logSearchWikipediaClick() {
+        action = SEARCH_WIKIPEDIA.valueString
+        submitEvent()
+    }
+
+    fun logBackClick() {
+        action = BACK.valueString
+        submitEvent()
+    }
+
+    fun logEditHistoryArticleClick() {
+        action = EDIT_HISTORY_ARTICLE.valueString
+        submitEvent()
+    }
+
+    fun logTalkPageArticleClick() {
+        action = TALK_PAGE_ARTICLE.valueString
+        submitEvent()
+    }
+
+    fun logTocSwipe() {
+        action = TOC_SWIPE.valueString
+        submitEvent()
+    }
+
+    fun logCategoriesClick() {
+        action = CATEGORIES.valueString
+        submitEvent()
+    }
+
+    private fun submitEvent() {
+        timeSpentMs = duration.toInt()
+        isAnon = !AccountUtil.isLoggedIn
+        EventPlatformClient.submit(this)
```

Comment:
This doesn't seem to follow the principles of OOP.  With every call of this `submitEvent()` function, you re-submit the same object (`this`) to be sent. But all the different `log...` functions _modify_ the `this` object. This means that if multiple references to `this` get queued up to be sent, they'll all contain the same information, even if they were trying to log different things.

Every item submitted into `EventPlatformClient` must be a unique instance of an event. For an example, see `UserContributionEvent`.

Reply:
@dbrant We had the exact opposite of this conversation in a pair programming session for this PR https://github.com/wikimedia/apps-android-wikipedia/pull/3040/commits and as you can see, in the first commit i had sent unique instances of the event, but then we decided that since we will never queue them up, we will just manipulate the object and keep re-sending.. so which way should we do this? Shall we decide once for all, that some day we will queue events and then have all MEP classes send unique instances? I have even made a ticket to change them all to send `this` after that conversation we had  https://phabricator.wikimedia.org/T299240

Reply:
So the series of decisions was  - we were not finding parity between legacy and MEP events - we made many other code tightening changes, but it remained the same - Then we decided to try the not queuing up solution to make it more of apples-to-apples  comparison, which in turn fixed how we were seeing the events. So the conclusion was that there must be something not fully flushed out at the MEP end about handling batched events [ speculation] as we were doing what was expected at our end. We should definitely discuss the path forward with batching. This makes sense #3240 , so will merge it and make similar changes for this schema as well.